platform-sdk/docs/proposals/queuestate-fcqueue-to-virtualmap/proposal.md

```diff
@@ -0,0 +1,159 @@
+# Design proposal: queue states - migrate from FCQueue to VirtualMap
+
+---
+
+## Summary
+
+This design proposal covers details how to re-implement Queue state APIs on top of `VirtualMap`
+rather than `FCQueue` as it is today. The main goal is to have applications like Hedera services
+unaffected, and have data migration fast enough to be done during software update cycle.
+
+## Purpose and context
+
+State APIs is an abstraction layer on top of the merkle state tree. This layer provides methods
+and classes to work with ledger data regardless of how it's stored in memory and on disk. There
+are 3 different state types supported:
+
+* Singleton states: store a single value
+* Queue states: store multiple values with queue-like APIs
+* Key/value states: store multiple values with map-like APIs
+
+For production use cases, queue states are implemented using a single `QueueNode` merkle node, which
+is an internal merkle node with two children: a label and a queue. The label is a `StringLeaf`, and
+the queue is an `FCQueue`.
+
+`FCQueue` is an in-memory implementation of a queue. It provides all APIs expected from a queue,
+like to add an element to the head of the queue or to poll an element at the tail of the queue. It
+also provides efficient implementation of hashCode(). This is critical, since queues may contain
+lots of elements, and queue hashes must be computed fast.
+
+There are several drawbacks in `FCQueue`, too:
+
+* All data is in memory
+* When a state snapshot is taken, all queue elements are serialized to disk one by one
+* During reconnects, the whole queue is serialized and sent to the learner
+
+Instead, all the queue APIs can be implemented on top of a `VirtualMap`. It's more memory friendly,
+optimizied for reconnects and snapshots.
+
+## Architecture
+
+### Diagrams
+
+|              Current              |                Proposal                |
+|-----------------------------------|----------------------------------------|
+| ![Current](architecture-0.54.svg) | ![Proposal](architecture-proposal.svg) |
+
+### Java classes
+
+No changes to `ReadableQueueState` and `WritableQueueState` interfaces.
+
+No changes to `ReadableQueueStateImpl` and `WritableQueueStateImpl` classes.
+
+`QueueNode`:
+
+* No longer marked as `@Deprecated`
+* Has two children: `QueueNodeState` (was: `StringLeaf`) and `VirtualMap` (was: `FCQueue`)
+
+`QueueNodeState`:
+
+* A simple leaf node similar to `StringLeaf`, but contains two long values in addition to
+  a string value
+* These two longs are the head and the tail of the queue - see below
```

Comment:
When the whole state tree is migrated to a mega map, tree structure will be completely different. There will be no state labels (those `StringLeaf`s above), no `SingletonNode` or `QueueNode`, and much more. This is beyond the scope of the current proposal. At this point, only queue state implementation (`QueueNode` and below) is changed, while the rest of the tree is unaffected

Reply:
How far off is mega-map? Is it worth doing the interim step?

Reply:
The process can be split the other way around: introduce the mega-map first, and then get back to refactoring the queues after the mega-map exists. I believe this would require less of throw-away work, so I disagree with your argument.

But if you insist on delivering the work in the current sequence, then please feel free to resolve this conversation.

Reply:
I don't see how it can be done the other way round. When a single virtual map is implemented, all data should be in it. All existing states (singletons, queues, maps) will be migrated to that single map. It means, there will be migration code for queues. There must be an answer to the question how to map queue elements to key/values. This is what covered in this design proposal. The idea is to have two smaller steps rather than a single larger one, but no code will be thrown away

Reply:
The code (and disk artifacts) that create and support a dedicated virtual map will be a throw away code. Similarly, the code that's going to use the map is going to have to migrate to using the States API constructs (or whatever API the mega-map will expose), so at the low level it will also be a throw away code.

Generally, the assumption was that the migration to the mega-map would be gradual - one or a few data structures at a time. Your last comment seems to suggest that it's going to be a big-bang with migrating everything at once. I find this approach risky, so again, I cannot fully agree with it.

But like I said, if you prefer to deliver this work this way, please feel free to resolve this conversation.